pkg/controller/node/node_controller.go

```diff
@@ -1022,13 +1036,53 @@ func (ctrl *Controller) setUpdateInProgressTaint(ctx context.Context, nodeName s
 			newNode.Spec.Taints = []corev1.Taint{}
 		}
 
+		if checkIfNodeHasInProgressTaint(newNode) {
+			return nil
+		}
+
+		newNode.Spec.Taints = append(newNode.Spec.Taints, *constants.NodeUpdateInProgressTaint)
+		newData, err := json.Marshal(newNode)
+		if err != nil {
+			return err
+		}
+
+		patchBytes, err := strategicpatch.CreateTwoWayMergePatch(oldData, newData, corev1.Node{})
+		if err != nil {
+			return fmt.Errorf("failed to create patch for node %q: %v", nodeName, err)
+		}
+		_, err = ctrl.kubeClient.CoreV1().Nodes().Patch(ctx, nodeName, types.StrategicMergePatchType, patchBytes, metav1.PatchOptions{})
+		return err
+	})
+}
+
+func (ctrl *Controller) removeUpdateInProgressTaint(ctx context.Context, nodeName string) error {
+	return clientretry.RetryOnConflict(constants.NodeUpdateBackoff, func() error {
+		oldNode, err := ctrl.kubeClient.CoreV1().Nodes().Get(ctx, nodeName, metav1.GetOptions{})
+		if err != nil {
+			return err
+		}
+		oldData, err := json.Marshal(oldNode)
+		if err != nil {
+			return err
+		}
+
+		newNode := oldNode.DeepCopy()
+
+		// New taints to be copied.
+		var taintsAfterUpgrade []corev1.Taint
 		for _, taint := range newNode.Spec.Taints {
 			if taint.MatchTaint(constants.NodeUpdateInProgressTaint) {
-				return nil
+				continue
+			} else {
+				taintsAfterUpgrade = append(taintsAfterUpgrade, taint)
 			}
 		}
-
-		newNode.Spec.Taints = append(newNode.Spec.Taints, *constants.NodeUpdateInProgressTaint)
+		// updateInProgress taint is not there, so no need to patch the node object, return immediately
+		if len(taintsAfterUpgrade) == len(newNode.Spec.Taints) {
```

Comment:
This is fine, but for consistency purposes, maybe we can add a check above for
```
if !checkIfNodeHasInProgressTaint(oldNode) {
	return nil
}
```
instead?